File: Timer.java
// Timer.java
// Кухонный таймер на Java с ScheduledExecutorService

import java.util.Scanner;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class Timer {
    private int seconds;
    private int remaining;
    private boolean paused;
    private boolean running;
    private ScheduledExecutorService executor;

    public Timer(int seconds) {
        this.seconds = seconds;
        this.remaining = seconds;
        this.paused = false;
        this.running = false;
    }

    public void start() {
        this.running = true;
        this.executor = Executors.newSingleThreadScheduledExecutor();
        this.executor.scheduleAtFixedRate(() -> {
            if (!paused && running) {
                remaining--;
                if (remaining <= 0) {
                    running = false;
                    executor.shutdown();
                    beep();
                    System.out.println("\n⏰ Время вышло!");
                    System.exit(0);
                }
            }
        }, 1, 1, TimeUnit.SECONDS);
        displayLoop();
    }

    private void displayLoop() {
        new Thread(() -> {
            while (running) {
                int m = Math.max(0, remaining) / 60;
                int s = Math.max(0, remaining) % 60;
                System.out.printf("\rОсталось: %02d:%02d   ", m, s);
                try {
                    Thread.sleep(200);
                } catch (InterruptedException e) {
                    break;
                }
            }
        }).start();
    }

    public void pause() {
        if (!paused && running) {
            paused = true;
            System.out.println("\n⏸  Пауза");
        }
    }

    public void resume() {
        if (paused) {
            paused = false;
            System.out.println("\n▶  Возобновлено");
        }
    }

    public void stop() {
        running = false;
        if (executor != null) {
            executor.shutdown();
        }
    }

    public void addMinute() {
        remaining += 60;
    }

    public void subtractMinute() {
        if (remaining >= 60) {
            remaining -= 60;
        }
    }

    private void beep() {
        System.out.print('\007');
    }

    private static int parseTime(String arg) throws NumberFormatException {
        if (arg.contains(":")) {
            String[] parts = arg.split(":");
            return Integer.parseInt(parts[0]) * 60 + Integer.parseInt(parts[1]);
        } else {
            return Integer.parseInt(arg);
        }
    }

    private static void showHelp() {
        System.out.println("Управление: p - пауза, r - возобновить, q - выход, + - +1 мин, - - -1 мин");
    }

    public static void main(String[] args) throws Exception {
        int seconds;
        if (args.length > 0) {
            seconds = parseTime(args[0]);
        } else {
            Scanner scanner = new Scanner(System.in);
            System.out.print("Введите время (сек или MM:SS): ");
            String input = scanner.nextLine().trim();
            seconds = parseTime(input);
        }

        Timer timer = new Timer(seconds);
        System.out.printf("\n⏳ Таймер запущен на %02d:%02d\n", seconds/60, seconds%60);
        showHelp();
        timer.start();

        // Управление с клавиатуры
        Scanner scanner = new Scanner(System.in);
        while (true) {
            char c = scanner.next().charAt(0);
            switch (c) {
                case 'p': case 'P': timer.pause(); break;
                case 'r': case 'R': timer.resume(); break;
                case 'q': case 'Q':
                    timer.stop();
                    System.out.println("\nТаймер остановлен.");
                    System.exit(0);
                    break;
                case '+': timer.addMinute(); break;
                case '-': timer.subtractMinute(); break;
            }
        }
    }
}

File: Timer.cs
// Timer.cs
// Кухонный таймер на C# с System.Timers.Timer

using System;
using System.Threading;

class Timer
{
    private int seconds;
    private int remaining;
    private bool paused;
    private bool running;
    private System.Timers.Timer timer;
    private Thread displayThread;

    public Timer(int seconds)
    {
        this.seconds = seconds;
        this.remaining = seconds;
        this.paused = false;
        this.running = false;
    }

    public void Start()
    {
        this.running = true;
        this.timer = new System.Timers.Timer(1000);
        this.timer.Elapsed += (sender, e) =>
        {
            if (!paused && running)
            {
                remaining--;
                if (remaining <= 0)
                {
                    running = false;
                    timer.Stop();
                    Beep();
                    Console.WriteLine("\n⏰ Время вышло!");
                    Environment.Exit(0);
                }
            }
        };
        this.timer.Start();
        DisplayLoop();
    }

    private void DisplayLoop()
    {
        this.displayThread = new Thread(() =>
        {
            while (running)
            {
                int m = Math.Max(0, remaining) / 60;
                int s = Math.Max(0, remaining) % 60;
                Console.Write($"\rОсталось: {m:00}:{s:00}   ");
                Thread.Sleep(200);
            }
        });
        this.displayThread.IsBackground = true;
        this.displayThread.Start();
    }

    public void Pause()
    {
        if (!paused && running)
        {
            paused = true;
            Console.WriteLine("\n⏸  Пауза");
        }
    }

    public void Resume()
    {
        if (paused)
        {
            paused = false;
            Console.WriteLine("\n▶  Возобновлено");
        }
    }

    public void Stop()
    {
        running = false;
        if (timer != null) timer.Stop();
        if (displayThread != null) displayThread.Abort();
    }

    public void AddMinute() => remaining += 60;
    public void SubtractMinute() { if (remaining >= 60) remaining -= 60; }

    private void Beep()
    {
        Console.Beep(1000, 500);
    }

    private static int ParseTime(string arg)
    {
        if (arg.Contains(":"))
        {
            var parts = arg.Split(':');
            return int.Parse(parts[0]) * 60 + int.Parse(parts[1]);
        }
        else
        {
            return int.Parse(arg);
        }
    }

    private static void ShowHelp()
    {
        Console.WriteLine("Управление: p - пауза, r - возобновить, q - выход, + - +1 мин, - - -1 мин");
    }

    static void Main(string[] args)
    {
        int seconds;
        if (args.Length > 0)
        {
            seconds = ParseTime(args[0]);
        }
        else
        {
            Console.Write("Введите время (сек или MM:SS): ");
            string input = Console.ReadLine().Trim();
            seconds = ParseTime(input);
        }

        var timer = new Timer(seconds);
        Console.WriteLine($"\n⏳ Таймер запущен на {seconds/60:00}:{seconds%60:00}");
        ShowHelp();
        timer.Start();

        while (true)
        {
            char c = Console.ReadKey(true).KeyChar;
            switch (c)
            {
                case 'p': case 'P': timer.Pause(); break;
                case 'r': case 'R': timer.Resume(); break;
                case 'q': case 'Q':
                    timer.Stop();
                    Console.WriteLine("\nТаймер остановлен.");
                    return;
                case '+': timer.AddMinute(); break;
                case '-': timer.SubtractMinute(); break;
            }
        }
    }
}
